File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self):
        self.child = {}
        self.isTerminal = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        def insertUtil(word: str, root: TrieNode) -> None:
            if len(word) == 0:
                root.isTerminal = True
                return
            key = word[0]
            if key not in root.child:
                root.child[key] = TrieNode()
            insertUtil(word[1:], root.child[key])
        insertUtil(word, self.root)

    def search(self, word: str) -> bool:
        def searchUtil(word: str, root: TrieNode) -> bool:
            if not len(word): return root.isTerminal
            key = word[0]
            if key not in root.child: return False
            return searchUtil(word[1:], root.child[key])
        return searchUtil(word, self.root)

    def startsWith(self, prefix: str) -> bool:
        def startsWithUtil(prefix: str, root: TrieNode) -> bool:
            if not len(prefix): return True
            key = prefix[0]
            if key not in root.child: return False
            return startsWithUtil(prefix[1:], root.child[key])
        return startsWithUtil(prefix, self.root)        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.java
class TrieNode {
    public TrieNode[] child;
    public boolean isTerminal;

    public TrieNode(){
        this.child = new TrieNode[26];
        this.isTerminal = false;
    }
}

class Trie {
    private TrieNode root;

    public Trie() {
        this.root = new TrieNode();        
    }
    
    private void insertUtil(String word, TrieNode root){
        if(word.length() == 0){
            root.isTerminal = true;
            return;
        }
        int key = word.charAt(0) - 'a';
        if(root.child[key] == null) root.child[key] = new TrieNode();
        this.insertUtil(word.substring(1), root.child[key]);
    }

    public void insert(String word) {
        this.insertUtil(word, this.root);
    }

    private boolean searchUtil(String word, TrieNode root){
        if(word.length() == 0) return root.isTerminal;
        int key = word.charAt(0) - 'a';
        if(root.child[key] == null) return false;
        return this.searchUtil(word.substring(1), root.child[key]);
    }
    
    public boolean search(String word) {
        return this.searchUtil(word, this.root);
    }
    
    private boolean startsWithUtil(String prefix, TrieNode root){
        if(prefix.length() == 0) return true;
        int key = prefix.charAt(0) - 'a';
        if(root.child[key] == null) return false;
        return this.startsWithUtil(prefix.substring(1), root.child[key]);
    }

    public boolean startsWith(String prefix) {
        return this.startsWithUtil(prefix, this.root);
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */